fix currentRun.compiles(Symbol)

mkeskells opened the pull request:
before GenBCode the owner of classes change, so the `enclosingTopLevelClass` misses the top level class (probably changed in delambify) 
this allows the `compiles` to check the `originalOwner`, which seems to work in my test cases

mkeskells commented:
@lrytz @adriaanm as discussed via email

mkeskells commented:
/rebuild

mkeskells commented:
an alternate implementation for compiles might be
```
def compiles(sym: Symbol): Boolean =compiledFiles contains sym.pos.source.file
```
not sure why this would not work, and it seems simpler and faster (not yet benchmarked though)

retronym commented:
I don't believe that comparing `SourceFile` will have the right semantics, as `SourceFile.equals` just compares file paths.

```
scala> import scala.tools.nsc._, scala.tools.nsc.settings._
import scala.tools.nsc._
import scala.tools.nsc.settings._

scala> val g = new Global(new Settings)
g: scala.tools.nsc.Global = scala.tools.nsc.Global@25a52a60

scala> g.settings.usejavacp.value = true
g.settings.usejavacp.value: Boolean = true

scala> val C_sym_1 = { val r = new g.Run; r.compileSources(file :: Nil); r1.units.toList.head.body.children.last}
C_sym_1: g.Tree =
class C extends Object {
  def <init>(): C = {
    C.super.<init>();
    ()
  }
}

scala> val C_sym_2 = { val r = new g.Run; r.compileSources(file :: Nil); r1.units.toList.head.body.children.last}
C_sym_2: g.Tree =
class C extends Object {
  def <init>(): C = {
    C.super.<init>();
    ()
  }
}

scala> C_sym_1.pos.source == C_sym_2.pos.source
res7: Boolean = true
```

mkeskells commented:
@retronym hadn't thought of repl...

should this PR be targeted at 2.12 or 2.13?

I would hope that we could target this at 2.12 as its needed for the GenBCode changes to correctly predict BType expansion

retronym commented:
Benchmarks [arriving](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=e12ac748cd%7Ccbd596339f%7Cc47ab156b1%7Cc929228698&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench@scalabench@). (I'm a bit cautious about all the extra map lookups, but let's see what the benchmarks say)

retronym commented:
I'd be happy for this to target 2.12

mkeskells commented:
discussing how to avoid the map lookups
maybe compiles should be 
```
def compiles(sym: Symbol): Boolean =
      if (sym == NoSymbol) false
      else if (symSource.contains(sym)) true
      else if (!compiledFiles.contains(sym.pos.source.file.path)) false
      else if (!sym.isTopLevel) {
        var parent = sym
        while(!parent.owner.isTopLevel) parent = parent.owner
        var originalOwner = parent.originalOwner
        while(!originalOwner.isTopLevel) {
          parent = originalOwner
          originalOwner = parent.originalOwner
        }
        compiles(parent.originalEnclosingTopLevelClassOrDummy)
      }
      else if (sym.isModuleClass) compiles(sym.sourceModule)
      else false

```
and remove the method from Symbols. 

still need answers on other owner walking APIs like associatedFile

mkeskells commented:
changed to target 2.12

mkeskells commented:
@lrytz @retronym still need answers on other owner walking APIs like associatedFile

lrytz commented:
The PR looks fine to me in its current form, the benchmarks seem flat / within noise. The commits could be squashed into one.

mkeskells committed:
rework Symbol enclosingTopLevelClassOrDummy
 run.compiles checks use the originalOwner not the owner.
The owner may get adjusted in lamdalift to skip the class that a symbol
was declared within, so the chain can direct to a class is
apparently not compiled

Also remove redundant repeated calls to enclosingTopLevelClass

mkeskells commented:
squashed

retronym reviewed (approved):
LGTM, too. Thanks!

Jasper-M commented:
I think this is assigned to the wrong milestone.
